Fullscreen Mode: Change W button to toggle wp-admin sidebar

shaunandrews opened the pull request:
Fixes #22178

This is a second attempt; The first (and a lot more background info) is #21121.

> With the recent move to make Fullscreen mode the default, I've heard from many people that they expect to be able to toggle the WordPress sidebar menu by clicking on the WordPress logo in the top left. This PR tries that:

![wp-admin-toggle](https://user-images.githubusercontent.com/191598/81342952-19ba2800-9082-11ea-9a9c-0baa29b98beb.gif)

Commit:
Revert "Revert "Fullscreen Mode: Change the "Back" button to toggle the sidebar (#21121)" (#21929)"

This reverts commit caf52c16304e375e8a28047f47d3b69c81c9aa68.

shaunandrews committed:
Moving the adminMenuToggle outside of the skeleton. This could be a bad idea.

MichaelArestad commented:
Using this feels solid. I only have minor feedback:

In Chromium (Chrome and Edge Mac), I noticed there is an outline style around the W when clicked. It does not show up in Firefox. It made me think I had to click within the blue border to close it.

<img width="168" alt="image" src="https://user-images.githubusercontent.com/1123119/81344220-98f42f80-9073-11ea-9960-4955698f337b.png">

The other minor weirdness is around the speed of it closing. When I click in to the editor, it closes automatically, which is great. It does feel like it takes a touch too long. The timing might be good to explore in a future issue/PR.

talldan commented:
I couldn't see a linked issue, and #22178 was reported, so I've linked this PR as closing that issue.

youknowriad commented:
Seems like there's  a coupe bugs here

<img width="161" alt="Capture d’écran 2020-05-08 à 3 52 45 PM" src="https://user-images.githubusercontent.com/272444/81418080-0a9bae80-9144-11ea-82ba-bd30b46d77f6.png">

The safari issue is back (white spacec)

<img width="61" alt="Capture d’écran 2020-05-08 à 3 52 55 PM" src="https://user-images.githubusercontent.com/272444/81418110-138c8000-9144-11ea-8066-eaae92fa0013.png">

I think this state was one of the reasons the revert was suggested, should we hide the "Collapse menu" button on Fullscreen?

<img width="65" alt="Capture d’écran 2020-05-08 à 3 53 06 PM" src="https://user-images.githubusercontent.com/272444/81418182-299a4080-9144-11ea-9449-8ee8c356537d.png">

Also from that state, when you click the "w" again, you get this.

---

I think one of the other reasons, the revert was suggested was whether the "W" should be centered or not when the panel is open and potentially add the site title there? cc @mtias

shaunandrews commented:
Thanks @youknowriad — some of these I knew, some I didn't.

> I think one of the other reasons, the revert was suggested was whether the "W" should be centered or not when the panel is open and potentially add the site title there?

I don't have a strong preference on the centering, but I also haven't heard it brought up before. 

Adding a site title was explored, but I still think its best left for another PR so we can figure out a good way to displaying titles without aggressive truncation.

There was mention of having to integrate the wp-admin color schemes to the W button, and the lack of a selected state for the button. I'm still working on both of those.

shaunandrews committed:
Fixing two Safari issues related to the #adminmenuback element

- The #adminmenuback element wouldn't render when transitioned on screen.
- The same element wouldn't _leave_ when transitioned off screen.

I fixed both of these by hiding the #adminmenuback div entirely, and then adjusting the height of the #adminmenuwrap element to ensure the background color stretches vertically.

jasmussen commented:
Thanks for your endurance, Shaun.

It feels very important to me that users can _middle-click to open in a new tab_ on the logo. For people who do this, they expect to be able to, and when they can't, it can erode a bit of trust. In the before-times, we would keep the `<a href="...">` intact, to make this work, but add an `onclick` handler with a `return false` to stop the URL navigation.

The focus style doesn't meet contrast ratio. You could either do the double-shot version we have in master, or change the focus style to be white there. 

<img width="178" alt="Screenshot 2020-05-11 at 09 16 19" src="https://user-images.githubusercontent.com/1204802/81534331-27183080-9368-11ea-8997-0916b2a9e643.png">

<img width="163" alt="Screenshot 2020-05-11 at 09 17 15" src="https://user-images.githubusercontent.com/1204802/81534358-38613d00-9368-11ea-9ff5-d9d4d4c11d87.png">

I'm happy to see you were able to get the sidebar scrolling on short viewports, or admins with a lot of menu items:

<img width="1198" alt="Screenshot 2020-05-11 at 09 18 29" src="https://user-images.githubusercontent.com/1204802/81534708-ca694580-9368-11ea-82f0-814d66664fa5.png">

It's unfortunate that we still have to have the double-scrollbar on the right to make this happen; the fact that we got rid of that double scrollbar with fullscreen was my personal favorite feature of that. But at least now it's consistent with non-fullscreen. If your neurons can think up a way to make the left navigation rail scroll independantly, as a separate PR to this, I would buy you a milkshake. No, make that two milkshakes.

shaunandrews committed:
Hide the body's scrollbar until the admin menu is shown.

shaunandrews committed:
Adding a new color to the base-styles for the sidebar color.

shaunandrews committed:
Adding a sidebarText value to the base-styles package.

shaunandrews committed:
Implementing the color schemes, and simplifying the focus state.

shaunandrews commented:
> It feels very important to me that users can middle-click to open in a new tab on the logo.

I haven't forgotten about this; Next on my list here.

> The focus style doesn't meet contrast ratio. You could either do the double-shot version we have in master, or change the focus style to be white there.

I added a second, white border similar to what `master` does.

> It's unfortunate that we still have to have the double-scrollbar on the right to make this happen; the fact that we got rid of that double scrollbar with fullscreen was my personal favorite feature of that. 

I pushed a (hopeful) fix for this; Basically, I forced `overflow: hidden;` when full screen mode is enabled, and the sidebar is hidden. This feels like a better experience, in that you'd only see the double-scrollbar when you've toggled the sidebar. Let me know what you think, @jasmussen

--

I managed to update the colors to respect the wp-admin color scheme settings, which required adding two new values to the `base-styles` package.

Along with the request to allow a middle-click, I'm still considering wether the "folded" or "collapsed" mode of the sidebar makes sense in this context.

jasmussen commented:
> I added a second, white border similar to what master does.

I think we should try just the white, this feels a little heavy:

![focus](https://user-images.githubusercontent.com/1204802/81908723-def84880-95c9-11ea-9777-1031a6a4817e.gif)

There's also a `$border-width-focus` you can use so it's consistent with all other focus, instead of the 4px picked here:

<img width="395" alt="Screenshot 2020-05-14 at 10 02 15" src="https://user-images.githubusercontent.com/1204802/81908834-0cdd8d00-95ca-11ea-98fa-87eeef47acbf.png">

I also wonder whether it's actually nice for the focus to encompass the entire button rather than try to frame just the logo. Like so:

![focus alt](https://user-images.githubusercontent.com/1204802/81909705-4236aa80-95cb-11ea-9253-0e4f3d024a95.gif)

That should translate to this:

```
    box-shadow: inset 0 0 0 $border-width-focus #23282e, inset 0 0 0 #{ $border-width-focus + $border-width } $white;
```

I noticed a weird focus trap of sorts, though:

![focus trap](https://user-images.githubusercontent.com/1204802/81909799-64302d00-95cb-11ea-8761-321c6431bcb4.gif)

In short, the tabbing context changes depending on whether the menu is open or not. If the menu is closed, you tab the top menu. If the menu is open, you tab through the menu. Both these seem fine, the problem is I can't shift tab backwards from the menu into the logo anymore. Not sure what the best practice is here. 

> I pushed a (hopeful) fix for this; Basically, I forced overflow: hidden; when full screen mode is enabled, and the sidebar is hidden. This feels like a better experience, in that you'd only see the double-scrollbar when you've toggled the sidebar. Let me know what you think, @jasmussen

![scrollbars](https://user-images.githubusercontent.com/1204802/81910002-b2453080-95cb-11ea-8ea2-45f768186356.gif)

This seems to work as intended!

Still not ideal, but I don't think it can ever become ideal until we are able to touch the left navigation side rail directly.

And that, I think, actually the biggest challenge here — it's bridging old and new, and the edgecases that pop up, including what you mention about folded cases and such, are really really difficult to address in a good way. I'm impressed that you're tackling it head-on!

shaunandrews committed:
Moving __experimentalMainDashboardButton from the skeleton into the layout.

shaunandrews committed:
Change focus outline to use variable and simplier shadow.

shaunandrews committed:
Hide the submenus without affecting the height of the sidebar.

ItsJonQ commented:
👋 Haiii!! I can help out with the JS/interaction side of things :)

@jasmussen Feedback/questions for you!

> It feels very important to me that users can middle-click to open in a new tab on the logo...

Gotcha. So allow for the W logo to work like a link.
It's currently a `<button />` element. We can change it to a link.
If we do, what should the URL be?

For example, when I edit a post, the URL is:

```
http://localhost:8889/wp-admin/post.php?post=1080&action=edit
```

> ...I can't shift tab backwards from the menu into the logo anymore...

We can accommodate this. Just a heads up, the solution will be hacky. It's already quite hacky, but it'll be hackier. There isn't a way around this as the (React) `AdminMenuToggle` component is doing things in non-React `wp-admin` land.

Thanks!

shaunandrews committed:
First pass at adding the adminMenuToggle to the site editor.

jasmussen commented:
> If we do, what should the URL be?

I would use the same heuristic as we have today, it has felt natural for me for the duration of using fullscreen. I'm not 100% on what that heuristic is, I think it's posts when you edit posts, pages when you edit pages, and probably cpts work as well. So to an extent, I think of this more as "restoring the current behavior".

> We can accommodate this. Just a heads up, the solution will be hacky. It's already quite hacky, but it'll be hackier. There isn't a way around this as the (React) AdminMenuToggle component is doing things in non-React wp-admin land.

Let's be liberal with code comments, and let's also take notes for how we could do this better. I hope that we'll sooner rather than later be able to revisit the actual side navigation rail to improve things there, and if/when we do it would be good to know what to refactor to make things better.

vindl commented on packages/edit-site/src/components/editor/index.js:
```diff
@@ -92,6 +97,9 @@ function Editor( { settings: _settings } ) {
 		<>
 			<EditorStyles styles={ settings.styles } />
 			<FullscreenMode isActive={ isFullscreenActive } />
+			<MainDashboardButton.Slot>
```
I tried testing this with instructions from https://github.com/WordPress/gutenberg/pull/22323 to verify if extensibility of the close button, and it doesn't seem to work anymore. The default button is not replaced.

I'm not sure what's causing this, and whether it's somehow related to moving this outside of editor skeleton. FWIW I don't think that it should be moved out of this area. With this setup it seems that everyone who's aiming to provide a custom button, will also have to provide styles to position it properly on top of the header, and I think that will be prone to breakage over time.

Commit:
Opens admin/post page in new tab if middle click or ctrl/cmd + click

Commit:
Jump focus from WP-Admin menu back to Gutenberg toolbar on shift+tab

ItsJonQ commented:
@jasmussen + @shaunandrews Haiii!

I just pushed some updates that handles the 2 enhanced interactions:

1. Middle clicking (or ctrl/cmd + click) the (W) logo to open the in a new tab
2. Pressing <kbd>Shift</kbd>+<kbd>Tab</kbd> on the first WP Admin menu item to jump focus back to the Gutenberg toolbar (W) logo.

I've added a bunch of comments that describe how it works, and acknowledging that it's perhaps not the best 😅 

Let me know how it feels!

Commit:
Update AdminMenuToggle to display show/hide based on menu open state

ItsJonQ commented:
<img width="321" alt="Screen Shot 2020-05-19 at 1 18 54 PM" src="https://user-images.githubusercontent.com/2322354/82357659-67fcee80-99d3-11ea-9798-6372e3388ad4.png">

Oo! Pushed another update. This one updates the (Tooltip) label of the AdminMenuToggle (W) link to say "Show" / "Hide" depending on the open state.

Previously, it always said "Show"

shaunandrews commented:
@ItsJonQ Love the updates; Thanks!

My only bit of feedback is that the `href` should point to `edit.php` with `addQueryArgs` for the current document's `post_type`, like the old close button used to do:

https://github.com/WordPress/gutenberg/pull/22191/files#diff-5e54b8ed48d99cb46a591a186b9ad138L37

I tried this locally, but for some reason I'm getting `undefined` for the `postType.slug`.

ItsJonQ commented:
@shaunandrews Thank you! I'm happy to change the URL of the `href`. I couldn't find it from the link you shared. Would you happen to have the code example to get that `edit.php` based URL from the old Button?

Thank you!

shaunandrews commented:
Sorry, the link highlights a deleted file. Its strange. I copy/pasted the old button code:

```function FullscreenModeClose() {
 	const { isActive, postType } = useSelect( ( select ) => {
 		const { getCurrentPostType } = select( 'core/editor' );
 		const { isFeatureActive } = select( 'core/edit-post' );
 		const { getPostType } = select( 'core' );

 		return {
 			isActive: isFeatureActive( 'fullscreenMode' ),
 			postType: getPostType( getCurrentPostType() ),
 		};
 	}, [] );

 	if ( ! isActive || ! postType ) {
 		return null;
 	}

 	return (
 		<Button
 			className="edit-post-fullscreen-mode-close"
 			icon={ wordpress }
 			iconSize={ 36 }
 			href={ addQueryArgs( 'edit.php', {
 				post_type: postType.slug,
 			} ) }
 			label={ get( postType, [ 'labels', 'view_items' ], __( 'Back' ) ) }
 		/>
 	);
 }```

ItsJonQ commented:
@shaunandrews Awesome! Thank you! I'll give it a shot

Commit:
Fix AdminMenuToggle link to post.php

ItsJonQ commented:
@shaunandrews I just pushed up a change. It looks like it's working for me 😊 . Would you be able to let me know?

(Also. I'll fix up that merge conflict)

shaunandrews commented:
> It looks like it's working for me

That's cause you did it a way better way than my failed attempt.

Something changed with the `tab` behavior; I can no longer use the keyboard to `tab` into the sidebar menu.

Commit:
Fix ref to buttonNode for click events

ItsJonQ commented:
> Something changed with the tab behavior; I can no longer use the keyboard to tab into the sidebar menu.

Ah yes, my bad! I broke the HTML element reference. I just pushed that fix, along with rebase with `master`.

paaljoachim commented:
Hey @shaunandrews and others can we get an update to how this PR is coming along.
Thanks.

Can we look at the possibility of getting this and associated issue/s into WordPress 5.5?

ItsJonQ commented:
Hmm! Looks like it has conflicts again. Will help resolve

Commit:
Merge branch 'master' into try/fullscreen-sidebar-toggle-2

moorscode commented on packages/interface/src/components/admin-menu-toggle/index.js:
```diff
@@ -0,0 +1,198 @@
+	};
+
+	const label = isOpen
+		? __( 'Hide sidebar menu' )
```
Technically this is a "sidebar menu", but isn't this the "admin navigation" of a users-site?
"Hide navigation" or "Hide admin navigation" would make more sense to me as a (non-technical) user.

ItsJonQ replied:
Thanks for the feedback! I'll make that update

Commit:
Update show/hide navigation label

mtias commented:
I want to sit on this one a bit more before we commit to adding it in time for 5.5. The main issue I see is in the case we are looking for the W menu to open page list / template list in the context of FSE or the page editor, and how it might end up being disruptive if we change the behaviour version to version.

shaunandrews commented:
Closing this for now.
